Docs: Improved tokenizer docs

clintongormley opened the pull request:
Added descriptions and runnable examples to all tokenizers

nik9000 commented on docs/reference/analysis/tokenizers.asciidoc:
```diff
@@ -1,34 +1,135 @@
-used to build <<analysis-custom-analyzer,custom analyzers>>.
+The tokenizer is also responsible for recording the order or _position_ of
+each term (used for phrase and word proximity queries) and the start and end
+_character offsets_ of the original word which the term represents (used foor
```
s/foor/for/

nik9000 commented on docs/reference/analysis/tokenizers.asciidoc:
```diff
@@ -1,34 +1,135 @@
+
+<<analysis-letter-tokenizer,Lowercase Tokenizer>>::
+
+The `lowercase` tokenizer, like the `letter` tokenizer,  divides text into
```
I'm kind of curious why we have this? It feels weird to have two ways to lowercase.

clintongormley replied:
It explains it on the `lowercase` tokenizer page: it performs better

nik9000 commented on docs/reference/analysis/tokenizers.asciidoc:
```diff
@@ -1,34 +1,135 @@
+<<analysis-keyword-tokenizer,Keyword Tokenizer>>::
+
+The `keyword` tokenizer is a ``noop'' tokenizer that accepts whatever text it
+is given and outputs the exact same text as a single term.  It can be combined
```
s/term/token/ ?

clintongormley replied:
I use term and token pretty much interchangeably.  Not sure if this is correct or not, or what the real distinction is

nik9000 commented on docs/reference/analysis/tokenizers.asciidoc:
```diff
@@ -1,34 +1,135 @@
+
+The `keyword` tokenizer is a ``noop'' tokenizer that accepts whatever text it
+is given and outputs the exact same text as a single term.  It can be combined
+with token filters to normalise output, e.g. lower-casing email addresses.
```
s/ to normalize output, e.g. lower-casing email addresses./ like lowercase to normalize the analyzed terms./ ?

nik9000 commented on docs/reference/analysis/tokenizers/classic-tokenizer.asciidoc:
```diff
@@ -1,19 +1,94 @@
 
-A tokenizer of type `classic` providing grammar based tokenizer that is
-a good tokenizer for English language documents. This tokenizer has 
+The `classic` tokenizer provides a grammar based tokenizer that is
```
s/provides a/is a/ ?

nik9000 commented on docs/reference/analysis/tokenizers/classic-tokenizer.asciidoc:
```diff
@@ -1,19 +1,94 @@
-A tokenizer of type `classic` providing grammar based tokenizer that is
-a good tokenizer for English language documents. This tokenizer has 
+The `classic` tokenizer provides a grammar based tokenizer that is
+a good tokenizer for English language documents. This tokenizer has
```
s/a good tokenizer for/good/ ?

nik9000 commented on docs/reference/analysis/tokenizers/classic-tokenizer.asciidoc:
```diff
@@ -1,19 +1,94 @@
+---------------------------
+[ The, 2, QUICK, Brown, Foxes, jumped, over, the, lazy, dog's, bone ]
+---------------------------
+
```
I wonder if we could do something with `// TESTRESPONSE` here. I mean, we don't need to give the user the full output of the _analyze API but it'd be nice to actually assert the result somehow.

nik9000 commented on docs/reference/analysis/tokenizers/edgengram-tokenizer.asciidoc:
```diff
@@ -1,80 +1,190 @@
 
-This tokenizer is very similar to `nGram` but only keeps n-grams which
-start at the beginning of a token.
+Edge N-Grams are useful for _search-as-you-type_ queries.
```
Maybe we should point people to the completion suggester here too?

clintongormley replied:
Done

nik9000 commented on docs/reference/analysis/tokenizers/edgengram-tokenizer.asciidoc:
```diff
@@ -1,80 +1,190 @@
+<2> The `autocomplete_search` analyzer searches for the terms `[quick, fo]`, both of which appear in the index.
+
+===========================================
 
```
It'd be nice to have a `// TESTRESPONSE` snippet here so we can be sure that the example works and so people see what is found. It'd be even nicer to have a document indexed that *won't* be found by the search but might be if the wrong search_analyzer is used.

nik9000 commented on docs/reference/analysis/tokenizers/pattern-tokenizer.asciidoc:
```diff
@@ -1,38 +1,151 @@
+The `pattern` tokenizer accepts the following parameters:
+
+[horizontal]
+[horizontal]
```
Repeat?

nik9000 commented on docs/reference/analysis/tokenizers/thai-tokenizer.asciidoc:
```diff
@@ -1,7 +1,36 @@
+<<analysis-standard-tokenizer,`standard` tokenizer>>.
+
+WARNING: This tokenizer may not be supported by all JREs. It is known to work
+with Sun/Oracle and Harmony JREs. If your application needs to be fully
```
Might want to mention OpenJDK too. Is Harmony JRE still a thing?

nik9000 commented on docs/reference/analysis/tokenizers/thai-tokenizer.asciidoc:
```diff
@@ -1,7 +1,36 @@
+
+WARNING: This tokenizer may not be supported by all JREs. It is known to work
+with Sun/Oracle and Harmony JREs. If your application needs to be fully
+portable, consider using  the ICUTokenizer instead.
```
Link for ICUTokenizer?

nik9000 commented:
Left some points about wording. I'd love to be able to test those results arrays. We could probably build something to make that work, given the number of _analyze results that we have floating around the in the docs. It is nice to have such a short response form in the docs.

nik9000 commented:
I think the "it'd be nice to test" part can come later. This already adds more tests by adding `// CONSOLE` snippets. I'd be happy to make a quick change to get the responses tested after this is merged.

clintongormley commented:
Thanks for the review @nik9000.  I addressed most of your comments in a9b3db65625b52664825ba2db18057f45c12a9c6

In e79d5a85afc6fde119c7ab58f04a1076a1456391 I added TESTRESPONSEs for all of the tokenizer examples, wrapping them in comment blocks so that they don't render in the docs, eg:

    /////////////////////

    [source,js]
    ----------------------------
    {  response body }
    ----------------------------
    // TESTRESPONSE

    /////////////////////

There's just one test in the edge ngrams which I'm not sure how to fix, as the `took` value will be different on every test run

clintongormley commented on docs/reference/analysis/tokenizers/edgengram-tokenizer.asciidoc:
```diff
@@ -1,80 +1,324 @@
+}
+-----------------------------------
+// CONSOLE
+// TEST[skip:The took value will change on every request]
```
This GET request will return a different `took` value every time.  I tried just removing it with with `TEST[s/"took".*"//]` but that didn't work.

nik9000 replied:
s/"took": 123/"took": "$body.took"/ is the way I've done it. Might be the
best way but it works.
On May 15, 2016 8:07 AM, "Clinton Gormley" <notifications@github.com> wrote:

> In docs/reference/analysis/tokenizers/edgengram-tokenizer.asciidoc
> <https://github.com/elastic/elasticsearch/pull/18356#discussion_r63290592>
> :
>
> > +POST my_index/_refresh
> > +
> > +GET my_index/_search
> > +{
> > +  "query": {
> > +    "match": {
> > +      "title": {
> > +        "query": "Quick Fo", <2>
> > +        "operator": "and"
> > +      }
> > +    }
> > +  }
> > +}
> > +-----------------------------------
> > +// CONSOLE
> > +// TEST[skip:The took value will change on every request]
>
> This GET request will return a different took value every time. I tried
> just removing it with with TEST[s/"took".*"//] but that didn't work.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/18356/files/e79d5a85afc6fde119c7ab58f04a1076a1456391#r63290592>
>

clintongormley commented:
I've added TESTRESPONSES for analyzers in https://github.com/elastic/elasticsearch/pull/18356/commits/38b89f9966fccf3433ed4b9b954bd49e9222ef34, and docs and TESTRESPONSES for character filters in https://github.com/elastic/elasticsearch/pull/18356/commits/89bbc7f861ea99c3af0f0db8718be64f7d62db79

The `pattern_replace` tests aren't working as `//TEST[continued]` isn't working as expected.

clintongormley commented on docs/reference/analysis/charfilters/pattern-replace-charfilter.asciidoc:
```diff
@@ -1,37 +1,296 @@
---------------------------------------------------
+----------------------------
+<1> Note the incorrect highlight.
+// TESTRESPONSE
```
I get the following exception when running the docs check:

    TESTRESPONSE not paired with a snippet at .../charfilters/pattern-replace-charfilter.asciidoc:296

nik9000 replied:
That means it didn't match the // TESTRESPONSE with the snippet above it.
It doesn't like if there is a newline between the snippet and the marker.
Or something like that.
On May 15, 2016 10:59 AM, "Clinton Gormley" <notifications@github.com>
wrote:

> In docs/reference/analysis/charfilters/pattern-replace-charfilter.asciidoc
> <https://github.com/elastic/elasticsearch/pull/18356#discussion_r63292994>
> :
>
> >  }
> > ---------------------------------------------------
> > +----------------------------
> > +<1> Note the incorrect highlight.
> > +// TESTRESPONSE
>
> I get the following exception when running the docs check:
>
> TESTRESPONSE not paired with a snippet at .../charfilters/pattern-replace-charfilter.asciidoc:296
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/18356/files/89bbc7f861ea99c3af0f0db8718be64f7d62db79#r63292994>
>

nik9000 replied:
Switch the <1> and the comment. Would you like to propose a more useful error message?

nik9000 commented on docs/reference/analysis/charfilters/pattern-replace-charfilter.asciidoc:
```diff
@@ -1,37 +1,296 @@
+  ]
+}
+----------------------------
+// TEST[continued]
```
A snippet can't be both `// TEST` and `// TESTRESPONSE`. They are (or, rather, they should be) mutually exclusive. This should just be `// TESTRESPONSE` because it isn't in sense format.

clintongormley commented:
Thanks @nik9000 - fixed that.  There are two issues which remain:

* The `replacement` of `"$1"` is interpreted as a stash variable and so fails.  This would fail in the REST tests too
* The `"took"` value is always different.  I tried `TESTRESPONSE[s/"took".*//] but it seems to remove this from the provided response snippet, rather than from the received response.

nik9000 commented:
> The replacement of "$1" is interpreted as a stash variable and so fails. This would fail in the REST tests too

`\\$1` I think will do it, but I'm not sure. If not then maybe disable that test and merge and I'll hack on the rest infrastructure to make it work?

> TESTRESPONSE[s/"took".*//]
`// TESTRESPONSE[s/"took".*/"took": "$body.took"/]

`s//` just changes the generated tests. "$body.took" gets it from the stash.

clintongormley commented:
> `// TESTRESPONSE[s/"took"./"took": "$body.took"/]

this worked, thanks

> \\$1 I think will do it, 

That won't work because it changes what is displayed/sent to ES during the test. It's skipped for now.

nik9000 commented on docs/reference/analysis/charfilters/pattern-replace-charfilter.asciidoc:
```diff
@@ -257,14 +257,14 @@ GET my_index/_search
 ----------------------------
 {
   "timed_out": false,
+  "took": $body.took,
```
I usually put a number here and use the replacement below to put the "$body.took" bit in. I usually use however long it took on my laptop so a particularly savvy ready gets a sense of how long it might take.

nik9000 commented:
> That won't work because it changes what is displayed/sent to ES during the test. It's skipped for now.

LGTM. If you merge I'll have a look at the escape thing.
